File: src/leetcode/medium/removeNthNodeFromEndOfList.ts
/**
    * desc: Given the head of a linked list, remove the nth node from the end of the list and return its head.

    Example 1: =>
    Input: head = [1,2,3,4,5], n = 2
    Output: [1,2,3,5]
    
    Example 2: =>
    Input: head = [1], n = 1
    Output: []
    
    Example 3: =>
    Input: head = [1,2], n = 1
    Output: [1]
    
    Constraints: =>
    The number of nodes in the list is sz.
    1 <= sz <= 30
    0 <= Node.val <= 100
    1 <= n <= sz
    
    Follow up: Could you do this in one pass?
 */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class ListNode {
	val: number;
	next: ListNode | null;
	constructor(val?: number, next?: ListNode | null) {
		this.val = val === undefined ? 0 : val;
		this.next = next === undefined ? null : next;
	}
}

const removeNthFromEnd = (
	head: ListNode | null,
	n: number
): ListNode | null => {
	let iterator: ListNode | null = head;
	let length: number = 0;

	while (iterator !== null) {
		length += 1;
		iterator = iterator.next;
	}

	iterator = head;

	let index: number = 0;

	while (iterator !== null) {
		if (length - n === 0) {
			return iterator.next;
		}

		index += 1;

		if (length - n === index) {
			iterator.next = iterator.next!.next;

			return head;
		}

		iterator = iterator.next;
	}

	return null;
};

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

const removeNthFromEndTwoPointer = (
	head: ListNode | null,
	n: number
): ListNode | null => {
	let fast: ListNode | null = head;
	let slow: ListNode | null = head;

	for (let i = 0; i < n; i++) {
		if (fast!.next === null) return head!.next;

		fast = fast!.next;
	}

	while (fast!.next !== null) {
		fast = fast!.next;
		slow = slow!.next;
	}

	slow!.next = slow!.next!.next;

	return head;
};

File: src/leetcode/medium/removeNthNodeFromEndOfList.py
# Given the head of a linked list, remove the nth node from the end of the list and return its head.

# Example 1: =>
# Input: head = [1,2,3,4,5], n = 2
# Output: [1,2,3,5]

# Example 2: =>
# Input: head = [1], n = 1
# Output: []

# Example 3: =>
# Input: head = [1,2], n = 1
# Output: [1]

# Constraints: =>
# The number of nodes in the list is sz.
# 1 <= sz <= 30
# 0 <= Node.val <= 100
# 1 <= n <= sz

# Follow up: Could you do this in one pass?

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        #  two passes
        curr_node = head

        length = 0
        while curr_node:
            length += 1
            curr_node = curr_node.next

        dummy_node = ListNode(-1)
        dummy_node.next = head
        curr_node = dummy_node

        while curr_node:
            length -= 1

            if length == n - 1:
                curr_node.next = curr_node.next.next
                return dummy_node.next

            curr_node = curr_node.next

        return dummy_node.next
